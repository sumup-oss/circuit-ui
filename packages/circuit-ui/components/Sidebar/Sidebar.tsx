/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Fragment, ReactNode } from 'react';
import { css } from '@emotion/react';

import styled, { StyleProps } from '../../styles/styled.js';

import { Aggregator } from './components/Aggregator/index.js';
import { Backdrop } from './components/Backdrop/index.js';
import { CloseButton } from './components/CloseButton/index.js';
import { Footer } from './components/Footer/index.js';
import { Header } from './components/Header/index.js';
import { NavItem } from './components/NavItem/index.js';
import { NavList } from './components/NavList/index.js';
import { Separator } from './components/Separator/index.js';

export interface SidebarProps {
  /**
   * The children component passed to the Sidebar
   */
  children: ReactNode;
  /**
   * Determines whether the Sidebar is open
   */
  open: boolean;
  /**
   * Accessibility label for the CloseButton
   */
  closeButtonLabel: string;
  /**
   * A function to handle the sidebar close
   */
  onClose: () => void;
}

const SIDEBAR_WIDTH = 256;

const baseStyles = ({ theme }: StyleProps) => css`
  display: flex;
  flex-direction: column;
  height: 100%;
  min-width: ${SIDEBAR_WIDTH}px;
  background-color: ${theme.colors.n900};
  transition: transform ${theme.transitions.default};
  position: absolute;
  transform: translateX(-100%);
  z-index: ${theme.zIndex.navigation};
  ${theme.mq.giga} {
    transform: translateX(0);
    position: relative;
  }
`;

const openStyles = ({ theme, isOpen }: StyleProps & { isOpen: boolean }) =>
  isOpen &&
  css`
    transform: translateX(0);
    ${theme.mq.giga} {
      transform: translateX(0);
    }
  `;

const Drawer = styled('nav')(baseStyles, openStyles);

/**
 * @legacy
 */
export function Sidebar({
  children,
  open,
  closeButtonLabel,
  onClose,
  ...props
}: SidebarProps) {
  return (
    <Fragment>
      <Drawer isOpen={open} {...props}>
        {children}
      </Drawer>
      <Backdrop
        visible={open}
        onClick={onClose}
        data-testid="sidebar-backdrop"
      />
      <CloseButton
        visible={open}
        label={closeButtonLabel}
        onClick={onClose}
        data-testid="sidebar-close-button"
      />
    </Fragment>
  );
}

/**
 * @legacy
 */
Sidebar.Header = Header;
/**
 * @legacy
 */
Sidebar.Footer = Footer;
/**
 * @legacy
 */
Sidebar.NavList = NavList;
/**
 * @legacy
 */
Sidebar.NavItem = NavItem;
/**
 * @legacy
 */
Sidebar.Aggregator = Aggregator;
/**
 * @legacy
 */
Sidebar.Separator = Separator;
