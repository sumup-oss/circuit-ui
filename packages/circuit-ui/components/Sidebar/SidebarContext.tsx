/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { createContext, ReactNode, useState } from 'react';

export const SidebarContext = createContext({
  isSidebarOpen: false,
  toggleSidebar: () => {},
});

export const SidebarContextConsumer = SidebarContext.Consumer;

export interface SidebarContextProviderProps {
  /**
   * The children received by SidebarContextProvider
   */
  children: ReactNode;
}

export function SidebarContextProvider({
  children,
}: SidebarContextProviderProps) {
  const [isSidebarOpen, setSidebarOpen] = useState(false);

  const toggleSidebar = () => {
    setSidebarOpen((prevState) => !prevState);
  };
  return (
    <SidebarContext.Provider
      value={{
        isSidebarOpen,
        toggleSidebar,
      }}
    >
      {children}
    </SidebarContext.Provider>
  );
}
