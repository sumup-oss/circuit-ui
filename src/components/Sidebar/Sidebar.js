/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import Header from './components/Header';
import Footer from './components/Footer';
import NavList from './components/NavList';
import NavItem from './components/NavItem';
import Backdrop from './components/Backdrop';
import CloseButton from './components/CloseButton';
import Aggregator from './components/Aggregator';
import Separator from './components/Separator';
import useClickHandler from '../../hooks/use-click-handler';

const SIDEBAR_WIDTH = 256;

const baseStyles = ({ theme }) => css`
  label: sidebar;
  display: flex;
  flex-direction: column;
  height: 100%;
  min-width: ${SIDEBAR_WIDTH}px;
  background-color: ${theme.colors.n900};
  transition: transform ${theme.transitions.default};
  position: absolute;
  transform: translateX(-100%);
  z-index: ${theme.zIndex.sidebar};
  ${theme.mq.giga} {
    transform: translateX(0);
    position: relative;
  }
`;

const openStyles = ({ theme, open }) =>
  open &&
  css`
    label: sidebar--open;
    transform: translateX(0);
    ${theme.mq.giga} {
      transform: translateX(0);
    }
  `;

const Drawer = styled('nav')(baseStyles, openStyles);

const Sidebar = ({
  children,
  open,
  closeButtonLabel,
  onClose,
  tracking,
  ...props
}) => {
  const handleClose = useClickHandler(onClose, tracking, 'sidebar-close');

  return (
    <Fragment>
      <Drawer open={open} {...props}>
        {children}
      </Drawer>
      <Backdrop
        visible={open}
        onClick={handleClose}
        data-testid="sidebar-backdrop"
      />
      <CloseButton
        visible={open}
        label={closeButtonLabel}
        onClick={handleClose}
        data-testid="sidebar-close-button"
      />
    </Fragment>
  );
};

Sidebar.propTypes = {
  /**
   * The children component passed to the Sidebar
   */
  children: PropTypes.node,
  /**
   * Tells if the Sidebar is open
   */
  open: PropTypes.bool,
  /**
   * Accessibility label for the CloseButton
   */
  closeButtonLabel: PropTypes.string.isRequired,
  /**
   * A function to handle the sidebar close
   */
  onClose: PropTypes.func,
  /**
   * Additional data that is dispatched with the close tracking event.
   */
  tracking: PropTypes.shape({
    label: PropTypes.string.isRequired,
    component: PropTypes.string,
    customParameters: PropTypes.object
  })
};

Sidebar.defaultProps = {
  children: '',
  open: false,
  onClose: null
};

Sidebar.Header = Header;
Sidebar.Footer = Footer;
Sidebar.NavList = NavList;
Sidebar.NavItem = NavItem;
Sidebar.Aggregator = Aggregator;
Sidebar.Separator = Separator;

export default Sidebar;
