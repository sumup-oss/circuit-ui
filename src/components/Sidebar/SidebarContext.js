/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { createContext } from 'react';
import PropTypes from 'prop-types';

const SidebarContext = createContext({
  isSidebarOpen: false,
  toggleSidebar: () => {}
});

const SidebarContextConsumer = SidebarContext.Consumer;

class SidebarContextProvider extends React.Component {
  state = { isSidebarOpen: false };

  toggleSidebar = () => {
    this.setState(prevState => ({ isSidebarOpen: !prevState.isSidebarOpen }));
  };

  render() {
    return (
      <SidebarContext.Provider
        value={{
          isSidebarOpen: this.state.isSidebarOpen,
          toggleSidebar: this.toggleSidebar
        }}
      >
        {this.props.children}
      </SidebarContext.Provider>
    );
  }
}

SidebarContextProvider.propTypes = {
  /**
   * The children received by SidebarContextProvider
   */
  children: PropTypes.node
};

export { SidebarContext, SidebarContextProvider, SidebarContextConsumer };
