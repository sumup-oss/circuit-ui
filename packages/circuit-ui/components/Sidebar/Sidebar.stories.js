/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { useState } from 'react';
import { css } from '@emotion/react';
import styled from '@emotion/styled';
import { Home, Profile, Sales, Shop } from '@sumup/icons';

import Button from '../Button';

import { Separator } from './components/Separator';

import Sidebar from '.';

export default {
  title: 'Navigation/Sidebar',
  component: Sidebar,
  subcomponents: {
    Header: Sidebar.Header,
    NavList: Sidebar.NavList,
    NavItem: Sidebar.NavItem,
    Aggregator: Sidebar.Aggregator,
    Separator: Sidebar.Separator,
    Footer: Sidebar.Footer,
  },
  parameters: {
    layout: 'fullscreen',
  },
};

const Viewport = styled.div`
  min-height: 500px;
`;

export const Base = (args) => {
  const [open, setOpen] = useState(true);
  const [selected, setSelected] = useState(1);

  return (
    <Viewport>
      {!open && (
        <Button
          onClick={() => setOpen(true)}
          css={css`
            position: absolute;
          `}
        >
          Open sidebar
        </Button>
      )}

      <Sidebar {...args} open={open} onClose={() => setOpen(false)}>
        <Sidebar.Header>Header</Sidebar.Header>
        <Sidebar.NavList>
          <Sidebar.NavItem
            key="home"
            label="Home"
            selected={selected === 1}
            onClick={() => setSelected(1)}
            defaultIcon={<Home />}
            selectedIcon={<Home />}
            tracking={{ label: 'trackingId-home' }}
          />
          <Sidebar.Aggregator
            key="list"
            selected={selected === 2}
            label="List"
            defaultIcon={<Sales />}
            selectedIcon={<Sales />}
          >
            <Sidebar.NavItem
              label="First"
              selected={selected === 21}
              onClick={() => setSelected(21)}
            />
            <Sidebar.NavItem
              label="Second"
              selected={selected === 22}
              onClick={() => setSelected(22)}
            />
            <Sidebar.NavItem
              label="Third"
              selected={selected === 23}
              onClick={() => setSelected(23)}
            />
          </Sidebar.Aggregator>
          <Sidebar.NavItem
            key="shop"
            label="Shop"
            disabled
            selected={selected === 3}
            defaultIcon={<Shop />}
            selectedIcon={<Shop />}
            onClick={() => setSelected(3)}
          />
          <Separator key="separator" />
          <Sidebar.NavItem
            key="me"
            label="Me"
            selected={selected === 4}
            defaultIcon={<Profile />}
            selectedIcon={<Profile />}
            onClick={() => setSelected(4)}
          />
        </Sidebar.NavList>
        <Sidebar.Footer>Footer</Sidebar.Footer>
      </Sidebar>
    </Viewport>
  );
};

Base.args = {
  closeButtonLabel: 'close-button',
};
