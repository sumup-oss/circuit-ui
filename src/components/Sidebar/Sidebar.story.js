/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useState } from 'react';
import styled from '@emotion/styled';
import { boolean } from '@storybook/addon-knobs/react';

import docs from './Sidebar.docs.mdx';
import Sidebar from '.';

import { ReactComponent as HomeEmpty } from './icons/home-empty.svg';
import { ReactComponent as ListEmpty } from './icons/list-empty.svg';
import { ReactComponent as MeEmpty } from './icons/me-empty.svg';
import { ReactComponent as HomeFull } from './icons/home-full.svg';
import { ReactComponent as ListFull } from './icons/list-full.svg';
import { ReactComponent as MeFull } from './icons/me-full.svg';

const Viewport = styled.div`
  height: 100vh;
`;

export default {
  title: 'Components|Sidebar',
  component: Sidebar,
  parameters: {
    docs: { page: docs },
    jest: ['Sidebar']
  }
};

const SidebarWithState = () => {
  const [selected, setSelected] = useState(1);

  return (
    <Viewport>
      <Sidebar open={true} onClose={() => null} closeButtonLabel="close-button">
        <Sidebar.Header>Header</Sidebar.Header>
        <Sidebar.NavList>
          <Sidebar.NavItem
            key="home"
            label="Home"
            selected={selected === 1}
            onClick={() => setSelected(1)}
            defaultIcon={<HomeEmpty />}
            selectedIcon={<HomeFull />}
          />
          <Sidebar.Aggregator
            key="list"
            selected={selected === 2}
            label="List"
            defaultIcon={<ListEmpty />}
            selectedIcon={<ListFull />}
          >
            <Sidebar.NavItem
              label={`First`}
              selected={selected === 4}
              onClick={() => setSelected(4)}
            />
            <Sidebar.NavItem
              label={`Second`}
              selected={selected === 5}
              onClick={() => setSelected(5)}
            />
            <Sidebar.NavItem
              label={`Third`}
              selected={selected === 6}
              onClick={() => setSelected(6)}
            />
          </Sidebar.Aggregator>
          <Sidebar.NavItem
            key="me"
            label="Me"
            disabled={boolean('Disabled item', false)}
            selected={selected === 3}
            defaultIcon={<MeEmpty />}
            selectedIcon={<MeFull />}
            onClick={() => setSelected(3)}
          />
        </Sidebar.NavList>
        <Sidebar.Footer>Footer</Sidebar.Footer>
      </Sidebar>
    </Viewport>
  );
};

export const base = () => <SidebarWithState />;
