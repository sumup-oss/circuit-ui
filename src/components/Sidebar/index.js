import Sidebar from './Sidebar';
import {
  SidebarContextProvider,
  SidebarContextConsumer
} from './SidebarContext';

export default Sidebar;
export { SidebarContextProvider, SidebarContextConsumer };
