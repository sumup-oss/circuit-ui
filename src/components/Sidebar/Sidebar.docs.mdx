import { Status, Props, Story } from '../../../.storybook/components';
import Sidebar from './Sidebar';

# Sidebar

<Status.InReview/><Status.Description>Under active development</Status.Description>

The sidebar is the primary navigational component on SumUp's web-based applications. It groups different
sections of the same product (the user dashboard, for example), allowing easy navigation between first
and the second level of navigations, keeping a strong anchor for our users to always know where they are.

<Story id="components-sidebar--base" />

<Props of={Sidebar} />

## When to use it

If you have a single product that is web-based and has multiple value propositions/actions for a given user,
you should use a sidebar to ensure the usability and navigation of that product.

The sidebar supports first and second level items. The first level is used for primary actions/value for our users within a product.
When you have related multiple sections or the breakdown of a certain value you should group them as secondary items under a single first level item,
which will act as an aggregator.

## Usage guidelines

- **Do** use sentence case for every single navigation items
- **Do** always position the sidebar on the left side on desktop and in the top side (collapsed version) on mobile
- **Do** make sure that the sidebar is always visible and sticky, both on desktop and mobile
- **Do** add icons to first level items to help our users understand what they should expect in that section
- **Do** group related items of a given action under a secondary
- **Do not** place any element or part of an element on the sidebar.
- **Do not** use the secondary level as anchors of the first level.

## Content guidelines

- **Do** use sentence case for naming navigation items
- **Do** order the items according to the priority/usage of the action/value
- **Do** place the log out (if your application has a logged in environment) as an item on the sidebar
- **Do not** use more than 3 words for your an item (aim for 1 word if possible)

## Usage in code

There are a couple of different components you'll need to construct your
Sidebar:

- `Sidebar` - Wrapper component, which must also be wrapped with a component
  that will determine the width of the expanded sidebar.
- `Sidebar.Header` - Header component
- `Sidebar.NavList` - Wrapper for the list of navigation items
- `Sidebar.Aggregator` - Wraps around a list of `Sidebar.NavItem` in case
  you want to provided nested navigation.
- `Sidebar.NavItem` - Individual navigation items.
- `Sidebar.Footer` - Footer component

### Adding items to the sidebar

When you want to create a sidebar with a flat navigation structure, you
can simply use something like this:

```jsx
import React, { useState } from 'react';
import { Sidebar } from '@sumup/circuit-ui';

const Navigation = () => {
  const [open, setOpen] = useState(false);
  const [selected, setSelected] = useState([false, false]);
  const handleSelect = newSelectedIndex => {
    setSelected(selected.map((s, i) => i === newSelectedIndex && !i);
  };

  return (
    <Sidebar open={open} onClose={() => setOpen(!open)} closeButtonLabel="close-button">
      <Sidebar.Header>Header</Sidebar.Header>
      <Sidebar.NavList>
        <Sidebar.NavItem
          selected={selected[0]}
          onClick={() => handleSelect(0)}
          label="Item #1"
        />
        <Sidebar.NavItem
          selected={selected[1]}
          onClick={() => handleSelect(1)}
          label="Item #2"
        />
      </Sidebar.NavList>
      <Sidebar.Footer>Footer</Sidebar.Footer>
    </Sidebar>
  );
};

```

### Adding submenus to the sidebar

In order to create submenus, you need to wrap the `Sidebar.NavItem` components
that you want to put inside a parent menu with the `Sidebar.Aggregator`
component. You can do that like this:

```jsx
<Sidebar.Aggregator label="Parent">
  <Sidebar.NavItem onClick={() => null} label="Child 1" selected={true} />
  <Sidebar.NavItem onClick={() => null} label="Child 2" selected={false} />
</Sidebar.Aggregator>
```

### Using the SidebarContext

To avoid keeping the logic of toggling the `Sidebar` on your application, you can use the available
`SidebarContextProvider` and `SidebarContextConsumer` that are implemented with the new version of
React's Context API. You can do that like this:

```jsx
import React, { useState } from 'react';
import { Sidebar, SidebarContextProvider, SidebarContextConsumer } from '@sumup/circuit-ui';

const Navigation = () => {
  const [selected, setSelected] = useState([false, false]);
  const handleSelect = newSelectedIndex => {
    setSelected(selected.map((s, i) => i === newSelectedIndex && !i);
  };

  return (
    <SidebarContextProvider>
      <SidebarContextConsumer>
        {({ isSidebarOpen, toggleSidebar }) => (
          <Sidebar
            open={isSidebarOpen}
            onClose={toggleSidebar}
            closeButtonLabel="close-button"
          >
            <Sidebar.Header>Header</Sidebar.Header>
            <Sidebar.NavList>
              <Sidebar.NavItem
                selected={selected[0]}
                onClick={() => handleSelect(0)}
                label="Item #1"
                />
              <Sidebar.NavItem
                selected={selected[1]}
                onClick={() => handleSelect(1)}
                label="Item #2"
              />
            </Sidebar.NavList>
            <Sidebar.Footer>Footer</Sidebar.Footer>
          </Sidebar>
        )}
      </SidebarContextConsumer
    </SidebarContextProvider>
  );
};

```

That way, you can use the `SidebarContextConsumer` to control the `Sidebar` open state on other
components of your application without the need to drill the toggling function to where you need
to use it.
