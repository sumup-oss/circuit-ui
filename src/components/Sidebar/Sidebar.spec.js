/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import Sidebar from './Sidebar';

describe('<Sidebar />', () => {
  it('should render and match the snapshot when closed', () => {
    const props = {
      open: false,
      onClose: jest.fn(),
    };
    const actual = create(<Sidebar {...props} />);
    expect(actual).toMatchSnapshot();
  });

  it('should render and match snapshot when open', () => {
    const props = {
      open: true,
      onClose: jest.fn(),
    };
    const actual = create(<Sidebar {...props} />);
    expect(actual).toMatchSnapshot();
  });

  it('should dispatch onClose when CloseButton is clicked', () => {
    const props = {
      open: true,
      onClose: jest.fn(),
    };
    const { getByTestId } = render(<Sidebar {...props} />);
    act(() => {
      fireEvent.click(getByTestId('sidebar-close-button'));
    });
    expect(props.onClose).toHaveBeenCalled();
  });

  it('should dispatch onClose when the Backdrop is clicked', () => {
    const props = {
      open: true,
      onClose: jest.fn(),
    };
    const { getByTestId } = render(<Sidebar {...props} />);
    act(() => {
      fireEvent.click(getByTestId('sidebar-backdrop'));
    });
    expect(props.onClose).toHaveBeenCalled();
  });

  describe('accessibility', () => {
    it('should meet accessibility guidelines', async () => {
      const wrapper = renderToHtml(<Sidebar closeButtonLabel="close" />);
      const actual = await axe(wrapper);
      expect(actual).toHaveNoViolations();
    });
  });
});
