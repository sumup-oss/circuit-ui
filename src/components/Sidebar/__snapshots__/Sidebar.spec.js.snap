// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render and match snapshot when open 1`] = `
HTMLCollection [
  .circuit-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  min-width: 256px;
  background-color: #212933;
  -webkit-transition: -webkit-transform 200ms ease-in-out;
  -webkit-transition: transform 200ms ease-in-out;
  transition: transform 200ms ease-in-out;
  position: absolute;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  z-index: 800;
  -webkit-transform: translateX(0);
  -ms-transform: translateX(0);
  transform: translateX(0);
}

@media (min-width:960px) {
  .circuit-0 {
    -webkit-transform: translateX(0);
    -ms-transform: translateX(0);
    transform: translateX(0);
    position: relative;
  }
}

@media (min-width:960px) {
  .circuit-0 {
    -webkit-transform: translateX(0);
    -ms-transform: translateX(0);
    transform: translateX(0);
  }
}

<nav
    class="circuit-0 circuit-1"
    open=""
  />,
  .circuit-0 {
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: #212933;
  -webkit-transition: opacity 200ms ease-in-out,visibility 200ms ease-in-out;
  transition: opacity 200ms ease-in-out,visibility 200ms ease-in-out;
  visibility: hidden;
  opacity: 0;
  z-index: 700;
  visibility: visible;
  opacity: 0.56;
}

@media (min-width:960px) {
  .circuit-0 {
    visibility: hidden;
  }
}

<div
    class="circuit-0 circuit-1"
    data-testid="sidebar-backdrop"
  />,
  .circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-radius: 5px;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
  border: 0;
  border-radius: 100%;
  position: absolute;
  bottom: 16px;
  right: 16px;
  -webkit-transition: opacity 150ms ease-in-out,visibility 150ms ease-in-out;
  transition: opacity 150ms ease-in-out,visibility 150ms ease-in-out;
  visibility: hidden;
  opacity: 0;
  z-index: 800;
  visibility: visible;
  opacity: 1;
}

.circuit-2:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-2:focus::-moz-focus-inner {
  border: 0;
}

.circuit-2:disabled,
.circuit-2[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-2:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-2:active {
  background-color: #EEF0F2;
  border-color: #5C656F;
}

@media (min-width:960px) {
  .circuit-2 {
    visibility: hidden;
  }
}

<button
    class="circuit-1 circuit-2"
    data-testid="sidebar-close-button"
    title="Close"
  >
    <svg
      fill="none"
      height="24"
      role="presentation"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 7l10 10m0-10L7 17"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="circuit-0"
    >
      Close
    </span>
  </button>,
]
`;

exports[`<Sidebar /> should render and match the snapshot when closed 1`] = `
HTMLCollection [
  .circuit-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100%;
  min-width: 256px;
  background-color: #212933;
  -webkit-transition: -webkit-transform 200ms ease-in-out;
  -webkit-transition: transform 200ms ease-in-out;
  transition: transform 200ms ease-in-out;
  position: absolute;
  -webkit-transform: translateX(-100%);
  -ms-transform: translateX(-100%);
  transform: translateX(-100%);
  z-index: 800;
}

@media (min-width:960px) {
  .circuit-0 {
    -webkit-transform: translateX(0);
    -ms-transform: translateX(0);
    transform: translateX(0);
    position: relative;
  }
}

<nav
    class="circuit-0 circuit-1"
  />,
  .circuit-0 {
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: #212933;
  -webkit-transition: opacity 200ms ease-in-out,visibility 200ms ease-in-out;
  transition: opacity 200ms ease-in-out,visibility 200ms ease-in-out;
  visibility: hidden;
  opacity: 0;
  z-index: 700;
}

@media (min-width:960px) {
  .circuit-0 {
    visibility: hidden;
  }
}

<div
    class="circuit-0 circuit-1"
    data-testid="sidebar-backdrop"
  />,
  .circuit-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: auto;
  height: auto;
  cursor: pointer;
  font-size: 16px;
  line-height: 24px;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: 700;
  border-radius: 5px;
  border-width: 1px;
  border-style: solid;
  -webkit-transition: opacity 200ms ease-in-out;
  transition: opacity 200ms ease-in-out;
  background-color: #FFFFFF;
  border-color: #9DA7B1;
  color: #323E49;
  padding: 8px calc(24px - 1px);
  border: 0;
  border-radius: 100%;
  position: absolute;
  bottom: 16px;
  right: 16px;
  -webkit-transition: opacity 150ms ease-in-out,visibility 150ms ease-in-out;
  transition: opacity 150ms ease-in-out,visibility 150ms ease-in-out;
  visibility: hidden;
  opacity: 0;
  z-index: 800;
}

.circuit-2:focus {
  outline: 0;
  box-shadow: 0 0 0 4px #AFD0FE;
}

.circuit-2:focus::-moz-focus-inner {
  border: 0;
}

.circuit-2:disabled,
.circuit-2[disabled] {
  opacity: 0.5;
  pointer-events: none;
  box-shadow: none;
}

.circuit-2:hover {
  background-color: #FAFBFC;
  border-color: #5C656F;
}

.circuit-2:active {
  background-color: #EEF0F2;
  border-color: #5C656F;
}

@media (min-width:960px) {
  .circuit-2 {
    visibility: hidden;
  }
}

.circuit-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

<button
    class="circuit-1 circuit-2"
    data-testid="sidebar-close-button"
    title="Close"
  >
    <svg
      fill="none"
      height="24"
      role="presentation"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 7l10 10m0-10L7 17"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-width="2"
      />
    </svg>
    <span
      class="circuit-0"
    >
      Close
    </span>
  </button>,
]
`;
